File: app/shared/src/commonMain/kotlin/com/tneff/cyppieagents/agentview/TranscriptFolding.kt
package com.tneff.cyppieagents.agentview

/**
 * Pure reducer that folds the raw event stream into the rendered transcript list.
 *
 * Kept side-effect-free and coroutine-free so the streaming/merge semantics can be unit-tested
 * directly. [AgentViewModel] applies it as events arrive.
 *
 * Rules:
 *  - [AgentEvent.AssistantText]: deltas sharing an `id` concatenate into one growing item; the
 *    latest `complete` flag wins. (Streaming text is one item, not N rows.)
 *  - [AgentEvent.ToolCall]: same `id` updates in place (e.g. RUNNING → OK/ERROR), keeping its
 *    position; a new `id` appends.
 *  - [AgentEvent.Result] / [AgentEvent.Notice]: append-only, but de-duplicated by `id` so a
 *    reconnect replay of the same terminal event does not double it.
 *
 * CYP-335 — an updated-in-place row keeps the [AgentEvent.tsMs] of the event that **created** it: the
 * displayed time answers "when did this happen", not "when was this row last touched". A tool call is
 * dated by its start, not by the result that resolves it minutes later.
 */
fun foldEvent(current: List<AgentEvent>, event: AgentEvent): List<AgentEvent> {
    val idx = current.indexOfFirst { it.id == event.id }
    return when (event) {
        is AgentEvent.AssistantText -> {
            val existing = current.getOrNull(idx) as? AgentEvent.AssistantText
            if (existing != null) {
                current.toMutableList().also {
                    // Copies from `existing`, so `tsMs` (and `id`) are the first delta's by construction.
                    it[idx] = existing.copy(
                        text = existing.text + event.text,
                        complete = event.complete,
                    )
                }
            } else {
                current + event
            }
        }

        is AgentEvent.ToolCall -> {
            val existing = current.getOrNull(idx) as? AgentEvent.ToolCall
            if (existing != null) {
                // Take the incoming payload (RUNNING → OK/ERROR) but keep the row's own start time: a plain
                // `it[idx] = event` re-dates the row to the moment the tool FINISHED.
                //
                // Precisely: an event coming through [StreamJsonMapper] already carries the start time here (its
                // `prior.copy(...)` preserves it), so for that source this is a second line of defence. It is the
                // ONLY line of defence for sources that do NOT go through the mapper — [StubAgentSession], tests,
                // and any future session emitting [AgentEvent]s directly. Not redundant; do not delete.
                current.toMutableList().also { it[idx] = event.copy(tsMs = existing.tsMs) }
            } else {
                current + event
            }
        }

        is AgentEvent.Result,
        is AgentEvent.Notice,
        // CYP-323: the human echo is append-only and de-duplicated by its stable turn id — idempotent
        // insurance if the same turn were ever re-applied (re-entrant send / a future Hub-echo path).
        is AgentEvent.UserTurn,
        // CYP-326 #1: the injected incoming/system message — append-only, de-duplicated by id (a reconnect
        // replay of the same synthetic event must not double it), rendered in stream order.
        is AgentEvent.IncomingSystem -> {
            if (idx >= 0) current else current + event
        }
    }
}

/** Folds a whole sequence from empty. Convenience for tests and replays. */
fun foldEvents(events: Iterable<AgentEvent>): List<AgentEvent> =
    events.fold(emptyList()) { acc, e -> foldEvent(acc, e) }

File: web-ts/src/agentview/transcriptFolding.ts
// CYP-401 (W3) — pure reducer folding the raw event stream into the rendered transcript, ported from
// :app:shared (agentview/TranscriptFolding.kt). Side-effect-free so the merge semantics are unit-tested directly.
//   - assistantText: deltas sharing an id concatenate into one growing row; the latest `complete` wins.
//   - toolCall: same id updates IN PLACE (RUNNING → OK/ERROR), keeping position AND the row's original tsMs
//     (CYP-335: a row is dated by when it happened, not when it was last touched).
//   - result / notice / userTurn / incomingSystem: append-only, de-duplicated by id (reconnect replay safe).
import type { AgentEvent, AssistantTextRow, ToolCallRow } from './agentEvent'

export function foldEvent(current: AgentEvent[], event: AgentEvent): AgentEvent[] {
  const idx = current.findIndex((e) => e.id === event.id)
  switch (event.kind) {
    case 'assistantText': {
      const existing = idx >= 0 && current[idx].kind === 'assistantText' ? (current[idx] as AssistantTextRow) : null
      if (existing !== null) {
        const next = [...current]
        // Spread from `existing`, so tsMs (and id) stay the FIRST delta's by construction.
        next[idx] = { ...existing, text: existing.text + event.text, complete: event.complete }
        return next
      }
      return [...current, event]
    }
    case 'toolCall': {
      const existing = idx >= 0 && current[idx].kind === 'toolCall' ? (current[idx] as ToolCallRow) : null
      if (existing !== null) {
        const next = [...current]
        // Take the incoming payload (RUNNING → OK/ERROR) but keep the row's own start time.
        next[idx] = { ...event, tsMs: existing.tsMs }
        return next
      }
      return [...current, event]
    }
    case 'result':
    case 'notice':
    case 'userTurn':
    case 'incomingSystem':
      return idx >= 0 ? current : [...current, event]
  }
}

/** Folds a whole sequence from empty — convenience for replays and the behavioral corpus. */
export function foldEvents(events: Iterable<AgentEvent>): AgentEvent[] {
  let acc: AgentEvent[] = []
  for (const e of events) acc = foldEvent(acc, e)
  return acc
}
